lfroms/COMP4109-Project
# Add end-to-end encryption for all sent messages

### What is the problem being solved in this PR?
This PR adds support for end-to-end encryption for all sent messages.  It also adds support for creating conversations between 2 hardcoded users.

### Related issues or PRs
Resolves #19

### How did you accomplish this and why did you choose this approach?
- By creating a `ConversationsList` component that is responsible for listing conversations.
- By creating a `Frame` component which renders the conversations list and wraps the entire router.
- By creating a `UserMenu` component which allows a user to log out.
- By modifying the `useConversation` hook so that it encrypts and decrypts messages as they are sent and received.
- By creating a `useConversations` hook used to manage conversations.  This hook will eventually be modified to be WebSocket-aware.
- By creating a `useRegistration` hook used to allow users to register on the app.
- By adding a `useSessionStorage` hook used to interact with SessionStorage.
- By creating a `useUserSession` hook used to interact with the user's current session.
- By adding a migration which adjusts the uniqueness constraints in the database.
- By modifying `[conversationId].tsx` so that it correctly handles encrypted messages.
- By adjusting all routes so that they return more consistent responses.
- By creating a `/register` route so that users may register using REST.
- By modifying `sendMessage` and `createConversation` socket events so that they better transmit encrypted payloads, and save them to the database.
- By adjusting the types to be more consistent.

### How to test changes
1. `psql postgres`
2. `\c chat_app`
2. `drop owned by admin;`
3. `exit`
4. `yarn install`
5. `yarn run typeorm migration:run`
6. `yarn start:dev`
7. Go to `localhost:3000`. Enter `lukas` (or your name) as the user id and press "Register".  After a few seconds, a `pem` file should be downloaded.  Rename it accordingly.
8. Go back to `localhost:3000`.  Enter `other` as the user id and press "Register". After a few seconds, a `pem` file should be downloaded.  Rename it accordingly.
9. Log in as user id `1` (enter this number in that text field).  Before pressing "Log in," click "Choose file" and upload the `pem` file for `lukas`.  Press "Log in".
10. Click "Create conversation"
11. Type and send a few messages.  Ensure that they appear as unencrypted text.
12. In an incognito window, Log in as user id `2` (enter this number in that text field).  Before pressing "Log in," click "Choose file" and upload the `pem` file for `other`.  Press "Log in".
13. Click the first link in the conversations list (this is the same conversation that `lukas` is participating in).
14. Send a few messages.  The first message may not appear, but subsequent messages should.  Messages should appear for both `lukas` and `other`.
15. Sigh of relief 🙂

### Checklist
